db/src/main/kotlin/edu/artic/db/AppDataManager.kt

```diff
@@ -1,38 +1,221 @@
+                            }
+                        }
+
+                        val mapAnnotations = result.mapAnnotations
+                        if (mapAnnotations?.isNotEmpty() == true) {
+                            try {
+                                appDatabase.articMapAnnotationDao.addAnnotations(mapAnnotations.values.toList())
+                            } catch (e: Throwable) {
+                                e.printStackTrace()
+                            }
+                        }
+
+                        try {
+                            appDatabase.dataObjectDao.setDataObject(result.data)
+                        } catch (e: Throwable) {
+                            e.printStackTrace()
+                        }
+                    }
+                    return@flatMap appDataState.asObservable()
+                }
+    }
+
+    private fun loadPostAppData(): Observable<Int> {
+        return Observable.create { observer ->
+            val maxDownloads = 2
+            var currentDownloads = 0
+            observer.onNext(currentDownloads)
+            getExhibitions().subscribe({
+                if (it is ProgressDataState.Done<*>) {
+                    currentDownloads++
+                    observer.onNext(currentDownloads)
+                }
+            }, {
+                observer.onError(it)
+            }, {
+                if (currentDownloads == maxDownloads) {
+                    observer.onComplete()
+                }
+            })
+            getEvents().subscribe({
+                if (it is ProgressDataState.Done<*>) {
+                    currentDownloads++
+                    observer.onNext(currentDownloads)
+                }
+            }, {
+                observer.onError(it)
+            }, {
+                if (currentDownloads == maxDownloads) {
+                    observer.onComplete()
+                }
+            })
+        }
+
+    }
+
+    private fun getExhibitions(): Observable<ProgressDataState> {
+        return serviceProvider.getExhibitions()
                 .flatMap {
-                    val dtf = DateTimeFormatterBuilder()
-                            //Thu, 12 Jul 2018 16:56:12 GMT
-                            .appendPattern("E, dd MMM yyyy HH:mm:ss O")
-                            .toFormatter()
-                    if (it.containsKey("last-modified")) {
-                        val lastModified = it["last-modified"]?.get(0)!!
-                        val newLocalDateTime = LocalDateTime.parse(lastModified, dtf)
-                        if(newLocalDateTime.isAfter(/*TODO: get from storage*/ LocalDateTime.parse("Thu, 12 Jul 2018 16:56:12 GMT", dtf))) {
-                            return@flatMap serviceProvider.getBlob()
-                        } else {
-                            return@flatMap Observable.just(AppDataState.Empty)
+                    when (it) {
+                        is ProgressDataState.Done<*> -> {
```

Comment:
maybe a generic extension filter method like `flatMapDone()` with the proper type casting would work here as it could cleanup code and reduce boilerplate.